crates/ra_hir/src/semantics.rs

```diff
@@ -0,0 +1,262 @@
+use std::{cell::RefCell, fmt, iter::successors};
+
+use ra_db::{FileId, FileRange};
+use ra_syntax::{ast, AstNode, SyntaxNode, SyntaxToken, TextRange};
+use rustc_hash::{FxHashMap, FxHashSet};
+
+use crate::{
+    db::HirDatabase, source_analyzer::ReferenceDescriptor, source_binder::ToDef, Function,
+    HirFileId, InFile, MacroDef, Module, Name, Origin, PathResolution, ScopeDef, SourceAnalyzer,
+    SourceBinder, StructField, Type, VariantDef,
+};
+use hir_def::TraitId;
+
+pub struct Semantics<'db, DB> {
+    pub db: &'db DB,
+    sb: RefCell<SourceBinder<'db, DB>>,
```

Comment:
I wish we could avoid `RefCell` for `SourceBinder` here...

Reply:
This is the perfect use-case for RefCell actually. Or, rather, it was an API mistake to not use `RefCell` in the `SourceBinder` in the first place. 

This type is not thread safe (as it stores syntax nodes) and in general is transient in nature. Additionally, what `RefCell` guards here are essentially caches. From the consumers pov, this looks like an immutable data structure. And `&T` APIs are much more convenient to use than `&mut T` APIs.

Reply:
So should we use `RefCell<FxHashMap<...>>` here inside of `SourceBinder` instead?https://github.com/rust-analyzer/rust-analyzer/blob/1fe48a0115c24240f5a3e1b329e642f18e2715d6/crates/ra_hir/src/source_binder.rs#L23-L26